crates/turborepo-filewatch/src/package_changes_watcher.rs

```diff
@@ -0,0 +1,226 @@
+            let Some(mut repo_state) = self.initialize_repo_state().await else {
+                return;
+            };
+            // TODO: Pass in global_deps and ignore_patterns
+            let mut change_mapper = ChangeMapper::new(&repo_state.pkg_dep_graph, vec![], vec![]);
+
+            loop {
+                match repo_state.file_events.recv().await {
+                    Ok(Ok(Event { paths, .. })) => {
+                        // No point in raising an error for an invalid .gitignore
+                        // This is slightly incorrect because we should also search for the
+                        // .gitignore files in the workspaces.
+                        let (root_gitignore, _) =
+                            Gitignore::new(&self.repo_root.join_component(".gitignore"));
+
+                        let changed_files: HashSet<_> = paths
+                            .into_iter()
+                            .filter_map(|p| {
+                                let p = AbsoluteSystemPathBuf::try_from(p).ok()?;
+                                self.repo_root.anchor(p).ok()
+                            })
+                            .filter(|p| !ancestors_is_ignored(&root_gitignore, p))
+                            .collect();
+
+                        let changes = change_mapper.changed_packages(changed_files.clone(), None);
+
+                        match changes {
+                            Ok(PackageChanges::All) => {
+                                // We tell the client that we need to rediscover the packages, i.e.
+                                // all bets are off, just re-run everything
+                                let _ = self
+                                    .package_change_events_tx
+                                    .send(PackageChangeEvent::Rediscover);
+                                match self.initialize_repo_state().await {
+                                    Some(new_repo_state) => {
+                                        repo_state = new_repo_state;
+                                        change_mapper = ChangeMapper::new(
+                                            &repo_state.pkg_dep_graph,
+                                            vec![],
+                                            vec![],
+                                        );
+                                    }
+                                    None => {
+                                        break;
+                                    }
+                                }
+                            }
+                            Ok(PackageChanges::Some(changed_pkgs)) => {
+                                tracing::debug!(
+                                    "changed files: {:?} changed packages: {:?}",
+                                    changed_files,
+                                    changed_pkgs
+                                );
+                                for pkg in changed_pkgs {
```

Comment:
Yeah I think in the client is maybe better since it keeps separation of concerns. Although perhaps if there's performance issues in the watcher, we could consider adding it there.